File: src/client/leo_bot.py
import os
import time
import Levenshtein

from bs4 import BeautifulSoup
from playwright.sync_api import Playwright
from global_config import config
from ai import ask

class LeoBot:
    def __init__(self):
        self.__playwright = None
        self.__browser = None
        self.__context = None
        self.__page = None
        
    def login(self, playwright: Playwright):
        """
        Выполняет вход в Telegram Web и авторизацию в leomatchbot.
        
        Args:
            playwright: Экземпляр Playwright для управления браузером
        """
        self.__playwright = playwright
        browser_context_path = config.get_browser_context_path()
        
        # Запускаем браузер (один раз для всего метода)
        self.__browser = self.__playwright.chromium.launch(headless=False)
        
        if not os.path.exists(browser_context_path):
            # Первый вход - требуется авторизация
            self._first_time_login(browser_context_path)
        else:
            # Используем сохраненную сессию
            self._restore_session(browser_context_path)
        
        """
        Переходит к leomatchbot через поиск.
        """
        print("🔍 Поиск leomatchbot...")
        
        # Кликаем на поле поиска и вводим имя бота
        self.__page.get_by_placeholder(" ").click()
        self.__page.get_by_placeholder(" ").fill("@leomatchbot")
        
        # Ожидаем появления бота в результатах поиска
        self.__page.wait_for_selector("text=leomatchbot", timeout=5000)
        
        # Переходим в чат с ботом
        self.__page.locator("text=leomatchbot").first.click()
        print("✅ Успешно перешли в чат с leomatchbot")

    def _first_time_login(self, context_path: str):
        """
        Выполняет первый вход с сохранением сессии.
        
        Args:
            context_path: Путь для сохранения контекста браузера
        """
        print("🔄 Первый запуск. Выполняется авторизация в Telegram Web...")
        
        # Создаем новый контекст
        self.__context = self.__browser.new_context()
        self.__page = self.__context.new_page()
        self.__page.goto("https://web.telegram.org/k/")
        
        # Ждем ручной авторизации пользователя
        self._wait_for_telegram_auth(self.__page)
        
        # Сохраняем состояние сессии
        self.__context.storage_state(path=context_path)
        print("✅ Сессия сохранена")

    def _restore_session(self, context_path: str):
        """
        Восстанавливает сессию из сохраненного контекста.
        
        Args:
            context_path: Путь к файлу с сохраненной сессией
        """
        print("🔄 Восстановление сохраненной сессии...")
        self.__context = self.__browser.new_context(storage_state=context_path)
        self.__page = self.__context.new_page()
        self.__page.goto("https://web.telegram.org/k/")

    def _wait_for_telegram_auth(self, page):
        """
        Ожидает ручной авторизации пользователя в Telegram Web.
        
        Args:
            page: Страница браузера с открытым Telegram Web
        """
        print("⏳ Ожидание авторизации в Telegram Web...")
        print("   Пожалуйста, войдите в свой аккаунт в открывшемся окне браузера")
        
        while True:
            try:
                time.sleep(1)
                page.get_by_placeholder(" ").click()
                page.get_by_placeholder(" ").fill("@leomatchbot")
                page.wait_for_selector("text=leomatchbot", timeout=5000)
                break
            except Exception as _:
                print("Не загрузился еще!")
                continue
    
    def get_page(self):
        return self.__page
    
    def get_content(self):
        return self.__page.content()

    def enter(self, message):
        """Умный поиск кнопки отправки"""
        
        # Попробовать разные стратегии по порядку
        selectors = [
            # 1. По data-атрибутам (если есть)
            "[data-qa='send-button']",
            "[data-test='send-btn']",
            "[aria-label*='send' i]",
            "[title*='send' i]",
            
            # 2. По классам
            ".send-button",
            ".btn-send",
            ".input-button",
            "button[type='submit']",
            
            # 3. По позиции относительно input
            ".input-message-input ~ button",
            ".input-message-input + button",
            
            # 4. По частичному тексту (если символы меняются)
            'button:has-text("")',  # стабильный символ
            'button:has-text("")',
            
            # 5. По XPath с contains
            'xpath=//button[contains(text(), "")]',
            
            # 6. Последняя кнопка в контейнере
            ".chat-footer button:last-child",
        ]
        
        for selector in selectors:
            button = self.__page.locator(selector).first
            if button.count() > 0:
                self.__page.locator(".input-message-input").first.click()
                self.__page.locator(".input-message-input").first.fill(message)
                button.click()
                return True
        
        # Если ничего не нашли - фоллбэк на координаты (нежелательно)
        self.__page.mouse.click(x=100, y=200)
        return False        
        
        
    def is_spam(self, text: str) -> bool:
        """
        Определяет, является ли текст спамом
        """
        req = f"""Ты - классификатор спама. Определи, является ли следующий текст спамом.
        
        Правила:
        1. Верни ТОЛЬКО одно слово: True или False
        2. True - если текст является спамом
        3. False - если текст не является спамом
        4. НЕ добавляй никаких пояснений, знаков препинания или дополнительного текста
        5. НЕ используй кавычки

        Текст для анализа: "{text}"

        Твой ответ (только True или False):
        """

        result = ask(req)
        
        # Очищаем результат от возможного мусора
        result = result.strip().lower()
        
        # Извлекаем True/False из ответа (на случай, если нейронка добавила лишнее)
        if 'true' in result:
            return True
        elif 'false' in result:
            return False
        else:
            # Если формат не распознан, логируем и возвращаем False по умолчанию
            print(f"Неожиданный формат ответа: {result}")
            return False
                
    
    def is_menu(self, text: str) -> bool:
        """
        1. Смотреть анкеты.\n2. Заполнить анкету заново.\n3. Изменить фото/видео.\n4. Изменить текст анкеты.
        
        
        Нашел 12053 девушек рядом с тобой. Показать?
        
        
        Ты понравился 1 девушке, показать её?
        
        1. Показать.
        2. Не хочу больше никого смотреть.
        """
        examples = [
            """
            1. Смотреть анкеты.
            2. Заполнить анкету заново.
            3. Изменить фото/видео.
            4. Изменить текст анкеты.
            """,
            """
            Несколько девушек из  хотят познакомиться с тобой прямо сейчас
            1. Посмотреть.
            2. Не интересно.
            """
        ]
        for example in examples:
            if self.__levenshtein_cmp(example, text, 0.6):
                return True
        return False
    
    def is_stop(self, text: str):
        """
        Слишком много ❤️ за сегодня.

        Пригласи друзей и получи больше ❤️!

        Перешли друзьям или размести в своих соцсетях.
        Вот твоя личная ссылка 👇
        """
        example = """
        Слишком много  за сегодня.
        Пригласи друзей и получи больше !
        Перешли друзьям или размести в своих соцсетях.
        Вот твоя личная ссылка\nБот знакомств Дайвинчик в Telegram! Найдет друзей или даже половинку 
        https://t.me/leomatchbot?start=
        Любые попытки купить или накрутить рефералов приводят к незамедлительной блокировке аккаунта.
        Кто-то тобой заинтересовался! Заканчивай с вопросом выше и посмотрим кто там'
        """
        return self.__levenshtein_cmp(example, text, 0.6)
    
    def is_match(self, text: str) -> bool:
        """
        Нашли кое-кого для тебя ;) Заканчивай с вопросом выше и увидишь кто это
        
        
        Ты понравился 1 девушке, показать её?

        1. Показать.
        2. Не хочу больше никого смотреть.
        
        """
        exampels = [
            """Нашли кое-кого для тебя ;) Заканчивай с вопросом выше и увидишь кто это""",
            """
            Ты понравился 1 девушке, показать её?
            1. Показать.
            2. Не хочу больше никого смотреть.
            """
        ]
        for example in exampels:
            if self.__levenshtein_cmp(example, text, 0.6) == True:
                return True
        return False


    def is_profile_url(self, text: str) -> bool:
        """
        Отлично! Надеюсь хорошо проведете время 🙌

        Начинай общаться 👉 Bunny
        """
        return False

    def download_image(self, image_url: str) -> bytes:
        """
        Скачивает одно изображение по blob URL и возвращает его как bytes без сохранения на диск
        """
        import base64
        
        try:
            # Используем существующую страницу для скачивания blob
            result = self.__page.evaluate("""
                (blobUrl) => {
                    return fetch(blobUrl)
                        .then(response => response.blob())
                        .then(blob => {
                            return new Promise((resolve) => {
                                const reader = new FileReader();
                                reader.onloadend = () => resolve(reader.result);
                                reader.readAsDataURL(blob);
                            });
                        })
                        .catch((error) => {
                            console.error('Fetch error:', error);
                            return null;
                        });
                }
            """, image_url)
            
            if result and result.startswith('data:'):
                # Извлекаем base64 данные
                header, encoded = result.split(',', 1)
                image_data = base64.b64decode(encoded)
                return image_data
            else:
                print(f"❌ Failed to download image: {image_url[:50]}...")
                return None
                
        except Exception as e:
            print(f"❌ Error downloading image: {e}")
            return None

    def extract_web_content(self) -> tuple[list, list]:
        soup = BeautifulSoup(self.__page.content(), 'html.parser')
        
        # 1. Находим контейнер с чатом
        chat_container = soup.find('div', class_='chat')
        if not chat_container:
            return None
        
        # 2. Ищем все сообщения в хронологическом порядке
        data = chat_container.find_all('div', attrs={'data-mid': True})
        
        text = []
        images = []
        for i, element in enumerate(data):
            text_elem = element.find('span', class_='translatable-message')
            if text_elem:
                text.append(text_elem.get_text().strip())
            
            media_imgs = element.find_all('img', class_='media-photo')
            for img in media_imgs:
                src = img.get('src')
                if src and src.startswith('blob:'):
                    images.append(src)
        
        return (text, images)

    def __levenshtein_cmp(self, text1: str, text2: str, threshold: float) -> bool:
            # Нормализуем тексты
        t1 = ' '.join(text1.lower().split())
        t2 = ' '.join(text2.lower().split())
        
        # Считаем расстояние
        distance = Levenshtein.distance(t1, t2)
        max_len = max(len(t1), len(t2))
        
        # Нормализованное сходство (1 - нормализованное расстояние)
        similarity = 1 - (distance / max_len) if max_len > 0 else 0
        
        return similarity >= threshold

File: src/client_rust/src/leo_bot.rs
use std::path::Path;
use std::time::Duration;
use std::fs;

use anyhow::{Result, anyhow};
use playwright::Playwright;
use playwright::api::{Browser, BrowserContext, Page};
use playwright::api::frame::{FrameState};
use scraper::{Html, Selector};
use tracing::{info, warn, error};
use tokio::time::sleep;
use fuzzywuzzy::fuzz;

use crate::config::CONFIG;
use leo_bot::utils::levenshtein_cmp;

pub struct LeoBot {
    playwright: Playwright,
    browser: Option<Browser>,
    context: Option<BrowserContext>,
    page: Option<Page>,
}

impl LeoBot {
    pub fn new(playwright: Playwright) -> Self {
        LeoBot {
            playwright: playwright,
            browser: None,
            context: None,
            page: None,
        }
    }
    
    pub async fn is_spam(&self, _text: &str) -> bool {
        false
    }
    
    pub async fn is_menu(&self, text: &str) -> bool {
        let examples = [
            r#"
            1. Смотреть анкеты.
            2. Заполнить анкету заново.
            3. Изменить фото/видео.
            4. Изменить текст анкеты.
            "#,
            r#"
            Несколько девушек из хотят познакомиться с тобой прямо сейчас
            1. Посмотреть.
            2. Не интересно.
            "#,
        ];
        
        for example in examples.iter() {
            if levenshtein_cmp(example, text, 0.6) {
                return true;
            }
        }
        
        false
    }
    
    pub async fn is_stop(&self, text: &str) -> bool {
        let example = r#"
        Слишком много за сегодня.
        Пригласи друзей и получи больше !
        Перешли друзьям или размести в своих соцсетях.
        Вот твоя личная ссылка
        Бот знакомств Дайвинчик в Telegram! Найдет друзей или даже половинку 
        https://t.me/leomatchbot?start= 
        Любые попытки купить или накрутить рефералов приводят к незамедлительной блокировке аккаунта.
        Кто-то тобой заинтересовался! Заканчивай с вопросом выше и посмотрим кто там
        "#;
        
        levenshtein_cmp(example, text, 0.6)
    }
    
    pub async fn is_match(&self, text: &str) -> bool {
        let examples = [
            r#"Нашли кое-кого для тебя ;) Заканчивай с вопросом выше и увидишь кто это"#,
            r#"
            Ты понравился 1 девушке, показать её?
            1. Показать.
            2. Не хочу больше никого смотреть.
            "#,
        ];
        
        for example in examples.iter() {
            if levenshtein_cmp(example, text, 0.6) {
                return true;
            }
        }
        
        false
    }

    pub async fn extract_web_content(&self) -> Result<(Vec<String>, Vec<String>)> {
        let page = self.page.as_ref().ok_or_else(|| anyhow!("Page not initialized"))?;
        
        let content = page.content().await?;
        let document = Html::parse_document(&content);
        
        let chat_selector = Selector::parse("div.chat").unwrap();
        let message_selector = Selector::parse("div[data-mid]").unwrap();
        let text_selector = Selector::parse("span.translatable-message").unwrap();
        let img_selector = Selector::parse("img.media-photo").unwrap();
        
        let mut texts = Vec::new();
        let mut images = Vec::new();
        
        if let Some(chat_container) = document.select(&chat_selector).next() {
            for element in chat_container.select(&message_selector) {
                if let Some(text_elem) = element.select(&text_selector).next() {
                    let text = text_elem.text().collect::<String>().trim().to_string();
                    if !text.is_empty() {
                        texts.push(text);
                    }
                }
                
                for img in element.select(&img_selector) {
                    if let Some(src) = img.value().attr("src") {
                        if src.starts_with("blob:") {
                            images.push(src.to_string());
                        }
                    }
                }
            }
        }
        
        Ok((texts, images))
    }

    pub async fn enter(&self, message: &str) -> Result<()> {
        let page = self.page.as_ref().ok_or_else(|| anyhow!("Page not initialized"))?;
        
        // Ищем поле ввода
        if let Ok(Some(input)) = page.query_selector(".input-message-input").await {
            input.click_builder().click().await?;
            input.fill_builder(message).fill().await?;
            
            // Ищем кнопку отправки
            let send_selectors = [
                "[data-qa='send-button']",
                "[data-test='send-btn']",
                ".send-button",
                ".btn-send",
                ".input-button",
                "button[type='submit']",
                ".chat-footer button:last-child",
            ];
            
            for selector in send_selectors.iter() {
                if let Ok(Some(button)) = page.query_selector(selector).await {
                    button.click_builder().click().await?;
                    return Ok(());
                }
            }
            
            input.press_builder("Enter")
                .press()
                .await?;
            
            return Ok(());
        }
        
        warn!("Поле ввода не найдено");
        Ok(())
    }

    pub async fn login(&mut self) -> Result<()> {
        let browser_context_path = CONFIG.get_browser_context_path().to_path_buf();
        let browser_path = CONFIG.get_chromium_path()
            .ok_or_else(|| anyhow!("Chromium not found in config"))?;

        let browser = self.playwright.chromium()
            .launcher()
            .headless(false)
            .executable(browser_path)
            .launch()
            .await?;

        if !browser_context_path.exists() {
            // Первый вход - требуется авторизация
            self.first_time_login(&browser, &browser_context_path).await?;
        } else {
            // Используем сохраненную сессию
            self.restore_session(&browser, &browser_context_path).await?;
        }

        info!("🔍 Поиск leomatchbot...");
        
        let page = self.page.as_ref().ok_or_else(|| anyhow!("Page not initialized"))?;
        
        // Ищем поле поиска и вводим текст
        let search_input = page.query_selector("[placeholder=\" \"]").await?
            .ok_or_else(|| anyhow!("Search input not found"))?;
        
        search_input.click_builder()
            .click()
            .await?;
        
        search_input.fill_builder("@leomatchbot")
            .fill()
            .await?;
        
        // Ждем появления бота - ИСПРАВЛЕНО
        page.wait_for_selector_builder("text=leomatchbot")
            .timeout(5000.0)
            .wait_for_selector()
            .await?;
        
        // Ищем бота и кликаем
        let bot_element = page.query_selector("text=leomatchbot").await?
            .ok_or_else(|| anyhow!("Bot element not found"))?;
        
        bot_element.click_builder()
            .click()
            .await?;
        
        info!("✅ Успешно перешли в чат с leomatchbot");

        self.browser = std::option::Option::Some(browser);
        Ok(())
    }

    async fn first_time_login(&mut self, browser: &Browser, context_path: &Path) -> Result<()> {
        info!("🔄 Первый запуск. Выполняется авторизация в Telegram Web...");
        
        // Создаем новый контекст
        let context = browser.context_builder().build().await?;
        let page = context.new_page().await?;
        
        page.goto_builder("https://web.telegram.org/k/")
            .goto()
            .await?;
        
        self.context = Some(context);
        self.page = Some(page.clone());
        
        // Ждем ручной авторизации пользователя
        self.wait_for_telegram_auth(&page).await?;
        
        // Сохраняем состояние сессии
        let storage_state = self.context.as_ref().unwrap().storage_state().await?;
        let json = serde_json::to_string_pretty(&storage_state)?;
        fs::write(context_path, json)?;
        
        info!("✅ Сессия сохранена");
        
        Ok(())
    }

    async fn restore_session(&mut self, browser: &Browser, context_path: &Path) -> Result<()> {
        info!("🔄 Восстановление сохраненной сессии...");
        
        // Читаем сохраненную сессию
        let json = fs::read_to_string(context_path)?;
        let storage_state = serde_json::from_str(&json)?;
        
        // Создаем контекст с восстановленной сессией
        let context = browser.context_builder()
            .storage_state(storage_state)
            .build()
            .await?;
            
        let page = context.new_page().await?;
        
        page.goto_builder("https://web.telegram.org/k/")
            .goto()
            .await?;
        
        self.context = Some(context);
        self.page = Some(page);
        
        info!("✅ Сессия восстановлена");
        
        Ok(())
    }

    async fn wait_for_telegram_auth(&self, page: &Page) -> Result<()> {
        info!("⏳ Ожидание авторизации в Telegram Web...");
        info!("   Пожалуйста, войдите в свой аккаунт в открывшемся окне браузера");
        
        loop {
            // Пробуем найти поле поиска
            if let Ok(Some(search_input)) = page.query_selector("[placeholder=\" \"]").await {
                // ИСПРАВЛЕНО: используем builder паттерн
                if let Err(e) = search_input.click_builder().click().await {
                    warn!("Ошибка при клике: {}", e);
                    continue;
                }
                
                if let Err(e) = search_input.fill_builder("@leomatchbot").fill().await {
                    warn!("Ошибка при вводе текста: {}", e);
                    continue;
                }
                
                // ИСПРАВЛЕНО: используем wait_for() метод
                match page.wait_for_selector_builder("text=leomatchbot")
                    .timeout(5000.0)
                    .wait_for_selector()
                    .await 
                {
                    Ok(_) => {
                        info!("✅ Бот найден, авторизация завершена");
                        return Ok(());
                    }
                    Err(e) => {
                        info!("Бот еще не появился, продолжаем ожидание... (ошибка: {})", e);
                        continue;
                    }
                }
            }
        }
        
        Ok(())
    }

    pub async fn download_image(&self, image_url: &str) -> Result<Vec<u8>> {
        let page = self.page.as_ref()
            .ok_or_else(|| anyhow!("Page not initialized"))?;
        
        let js_code = r#"
            (blobUrl) => {
                return fetch(blobUrl)
                    .then(response => response.blob())
                    .then(blob => {
                        return new Promise((resolve) => {
                            const reader = new FileReader();
                            reader.onloadend = () => resolve(reader.result);
                            reader.readAsDataURL(blob);
                        });
                    })
                    .catch((error) => {
                        console.error('Fetch error:', error);
                        return null;
                    });
            }
        "#;
        
        let result: Option<String> = page.evaluate(js_code, Some(serde_json::json!([image_url]))).await?;
        
        match result {
            Some(data_url) if data_url.starts_with("data:") => {
                if let Some(comma_idx) = data_url.find(',') {
                    let encoded = &data_url[comma_idx + 1..];
                    let decoded = BASE64.decode(encoded)?;
                    Ok(decoded)
                } else {
                    Err(anyhow!("Invalid data URL format"))
                }
            }
            Some(_) => {
                error!("❌ Failed to download image: {}...", &image_url.chars().take(50).collect::<String>());
                Err(anyhow!("Failed to download image: invalid response"))
            }
            None => {
                error!("❌ Failed to download image: {}...", &image_url.chars().take(50).collect::<String>());
                Err(anyhow!("Failed to download image: no data"))
            }
        }
    }
}